framework/src/utils/SlepcSupport.C

```diff
@@ -247,6 +273,95 @@ slepcSetOptions(FEProblemBase & problem)
   Moose::PetscSupport::addPetscOptionsFromCommandline();
 }
 
+void
+moosePetscSNESFormJacobian(
+    SNES /*snes*/, Vec x, Mat jac, Mat pc, void * ctx, Moose::KernelType type)
+{
+  EigenProblem * eigen_problem = static_cast<EigenProblem *>(ctx);
+  NonlinearSystemBase & nl = eigen_problem->getNonlinearSystemBase();
+  System & sys = nl.system();
+
+  PetscVector<Number> & X_sys = *cast_ptr<PetscVector<Number> *>(sys.solution.get());
+  PetscVector<Number> X_global(x, sys.comm());
+
+  // update local solution
+  X_global.swap(X_sys);
+  sys.update();
+  X_global.swap(X_sys);
+
+  PetscMatrix<Number> PC(pc, sys.comm());
+  PetscMatrix<Number> Jac(jac, sys.comm());
+
+  // Set the dof maps
+  PC.attach_dof_map(sys.get_dof_map());
+  Jac.attach_dof_map(sys.get_dof_map());
+
+  PC.zero();
+
+  eigen_problem->computeJacobian(*sys.current_local_solution.get(), PC, type);
+
+  PC.close();
+  if (jac != pc)
+    Jac.close();
+}
+
+PetscErrorCode
+mooseSlepcEigenFormJacobianA(SNES snes, Vec x, Mat jac, Mat pc, void * ctx)
+{
+  PetscFunctionBegin;
+
+  moosePetscSNESFormJacobian(snes, x, jac, pc, ctx, Moose::KT_NONEIGEN);
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode
+mooseSlepcEigenFormJacobianB(SNES snes, Vec x, Mat jac, Mat pc, void * ctx)
+{
+  PetscFunctionBegin;
+
+  moosePetscSNESFormJacobian(snes, x, jac, pc, ctx, Moose::KT_EIGEN);
+  PetscFunctionReturn(0);
+}
+
+void
+moosePetscSNESFormFunction(SNES /*snes*/, Vec x, Vec r, void * ctx, Moose::KernelType type)
+{
+  EigenProblem * eigen_problem = static_cast<EigenProblem *>(ctx);
+  NonlinearSystemBase & nl = eigen_problem->getNonlinearSystemBase();
+  System & sys = nl.system();
+
+  PetscVector<Number> & X_sys = *cast_ptr<PetscVector<Number> *>(sys.solution.get());
+  PetscVector<Number> X_global(x, sys.comm()), R(r, sys.comm());
+
+  // update local solution
+  X_global.swap(X_sys);
```

Comment:
I actually just copy this piece of code from libMesh. I guess they might have reason to do in this way. Before I change the code, I want comments from @jwpeterson.